File: internal/model/channel.go
package model

// 渠道支持的上游线协议, 以位掩码存储, 一条渠道授权可同时支持多个协议。
type Protocol uint8

// 协议位从 1 << 1 开始, 1 << 0 保留待用; 位值会落库并与前端交互, 不可再变更。
const (
	ProtocolOpenAIChatCompletion Protocol = 1 << 1 // OpenAI Chat Completions 协议。
	ProtocolOpenAIResponse       Protocol = 1 << 2 // OpenAI Responses 协议。
	ProtocolAnthropicMessage     Protocol = 1 << 3 // Anthropic Messages 协议。
	// ProtocolOpenAIImage 是 OpenAI 图片生成与编辑协议, 覆盖 /v1/images/generations 与 /v1/images/edits。
	// 两个端点共用一个协议位: 它们收发同一族请求体与响应体, 上游要么都提供要么都不提供,
	// 分成两位只会让授权矩阵多一列而没有能分别取舍的场景。
	ProtocolOpenAIImage Protocol = 1 << 4
)

// 上游在标准协议之上的方言。
// 同一协议下不同服务商仍可能有请求体或响应体上的差异, 例如思考内容放 reasoning_content 还是 reasoning,
// 这类差异无法由地址与路径表达, 需按方言分支构造出站转换器。
// 地址与路径不属于方言范畴, 由前端按服务商预填到渠道字段上。
type Dialect string

const (
	DialectGeneric Dialect = "generic" // 标准协议, 不做任何厂商特化。
)

// 渠道的可编辑配置; 落库时平铺成 channels 的各列, 出入 JSON 时平铺成渠道读写接口的各字段。
// 单独成结构体是为了让库内实体, 读取副本与提交请求共用同一份定义: 这些字段在三处的形状本就一致,
// 各写一遍只会让加字段时漏改其中一处。
// 不带 binding 约束: 保存与探测都收这一份配置, 但两者的必填项不同 —— 探测发生在渠道尚未命名时,
// 故必填校验分别由 normalizeChannelConfig 与 fetchUpstreamModels 按各自的需要给出。
type ChannelConfig struct {
	Name                      string         `json:"name" gorm:"unique;not null"`                                                                            // 渠道名称。
	Dialect                   Dialect        `json:"dialect" gorm:"not null;default:generic"`                                                                // 上游方言, 决定出站转换器的厂商特化配置。
	Enabled                   bool           `json:"enabled" gorm:"default:true"`                                                                            // 渠道是否可用。
	BaseURL                   string         `json:"base_url"`                                                                                               // 上游地址, 各协议共用。
	OpenAIChatCompletionPath  string         `json:"openai_chat_completion_path" gorm:"column:openai_chat_completion_path;default:/v1/chat/completions"`     // OpenAI Chat Completions 请求路径; 留空由后端填默认路径。
	OpenAIResponsePath        string         `json:"openai_response_path" gorm:"column:openai_response_path;default:/v1/responses"`                          // OpenAI Responses 请求路径; 留空由后端填默认路径。
	AnthropicMessagePath      string         `json:"anthropic_message_path" gorm:"column:anthropic_message_path;default:/v1/messages"`                       // Anthropic Messages 请求路径; 留空由后端填默认路径。
	OpenAIImageGenerationPath string         `json:"openai_image_generation_path" gorm:"column:openai_image_generation_path;default:/v1/images/generations"` // OpenAI 图片生成请求路径; 留空由后端填默认路径。
	OpenAIImageEditPath       string         `json:"openai_image_edit_path" gorm:"column:openai_image_edit_path;default:/v1/images/edits"`                   // OpenAI 图片编辑请求路径; 留空由后端填默认路径。
	Proxy                     bool           `json:"proxy" gorm:"default:false"`                                                                             // 是否使用代理。
	ChannelProxy              string         `json:"channel_proxy"`                                                                                          // 渠道专用代理地址; 留空表示不用渠道专用代理。
	CustomHeader              []CustomHeader `json:"custom_header" gorm:"serializer:json"`                                                                   // 追加到上游请求的 Header。
	ParamOverride             string         `json:"param_override"`                                                                                         // 请求参数覆盖配置; 留空表示不覆盖。
	MatchRegex                string         `json:"match_regex"`                                                                                            // 拉取模型列表时的过滤表达式; 留空表示不过滤。
}

// 单个上游渠道的共享配置; 路径按协议分别配置, 凭据由 ChannelKey 提供。
type Channel struct {
	ID            int            `json:"id" gorm:"primaryKey"` // 渠道主键。
	ChannelConfig                // 可编辑配置, 平铺为 channels 的各列。
	Keys          []ChannelKey   `json:"-" gorm:"foreignKey:ChannelID;constraint:OnDelete:CASCADE"` // 渠道下的上游凭据; 不出 JSON, 读取走 ChannelDetail。
	Models        []ChannelModel `json:"-" gorm:"foreignKey:ChannelID;constraint:OnDelete:CASCADE"` // 渠道提供的模型; 不出 JSON, 读取走 ChannelDetail。
	StatsMetrics                 // 渠道自身的累计统计。
}

// 渠道凭据的可编辑配置; 名称在渠道内唯一, 整体替换时以它为匹配依据。
// 与 ChannelConfig 同理, 库内实体与读写接口共用这一份定义。
type ChannelKeyConfig struct {
	Name    string `json:"name" gorm:"not null;index:idx_channel_key,unique"` // 凭据名称, 界面展示与人工识别用。
	Key     string `json:"key" gorm:"not null"`                               // 上游访问凭据。
	Enabled bool   `json:"enabled" gorm:"default:true"`                       // 是否可用, 禁用后不参与选路但保留统计。
}

// 渠道下的一份上游凭据; 不同凭据通常对应不同的额度与计费。
type ChannelKey struct {
	ID               int `json:"id" gorm:"primaryKey"`                                    // 凭据主键。
	ChannelID        int `json:"channel_id" gorm:"not null;index:idx_channel_key,unique"` // 所属渠道 ID。
	ChannelKeyConfig     // 可编辑配置, 平铺为 channel_keys 的各列。
	StatsMetrics         // 该凭据自身的累计统计。
}

// 渠道提供的单个上游模型。
type ChannelModel struct {
	ID           int    `json:"id" gorm:"primaryKey"`                                      // 渠道模型主键。
	ChannelID    int    `json:"channel_id" gorm:"not null;index:idx_channel_model,unique"` // 所属渠道 ID。
	Name         string `json:"name" gorm:"not null;index:idx_channel_model,unique"`       // 上游模型名称。
	StatsMetrics        // 该模型自身的累计统计。
}

// 渠道内的一条上游授权: 指定模型使用指定凭据时支持的协议集合, 也是转发的最小单位。
// 上游按凭据分组授权模型, 且同一凭据下不同模型可能位于不同协议的端点, 故协议归属于模型与凭据的组合。
type ChannelGrant struct {
	ID             int           `json:"id" gorm:"primaryKey"`                                                                               // 授权主键。
	ChannelModelID int           `json:"channel_model_id" gorm:"not null;index:idx_channel_grant,unique"`                                    // 渠道模型 ID。
	ChannelModel   *ChannelModel `json:"channel_model,omitempty" gorm:"foreignKey:ChannelModelID;references:ID;constraint:OnDelete:CASCADE"` // 授权引用的渠道模型。
	ChannelKeyID   int           `json:"channel_key_id" gorm:"not null;index:idx_channel_grant,unique"`                                      // 凭据 ID。
	ChannelKey     *ChannelKey   `json:"channel_key,omitempty" gorm:"foreignKey:ChannelKeyID;references:ID;constraint:OnDelete:CASCADE"`     // 授权引用的凭据。
	Protocols      Protocol      `json:"protocols" gorm:"not null"`                                                                          // 该组合支持的协议位掩码。
}

// 渠道读写副本: 编辑表单的完整形状, 读取与提交同构。
// 读写同构使前端无需为提交另建一套类型, 后端也无需按字段比对增量: 表单本就一次给出完整配置。
// 凭据与模型只给界面用得上的字段: 两者在渠道内按名称唯一, 提交时也按名称引用, 主键与统计都无从使用。
// 集合字段恒为数组, 读取侧承诺不为 null。
type ChannelDetail struct {
	ID            int                  `json:"id"` // 渠道主键; 创建时提交 0, 由数据库分配。
	ChannelConfig                      // 渠道自身的可编辑配置。
	Keys          []ChannelKeyConfig   `json:"keys"`   // 渠道下的上游凭据。
	Models        []string             `json:"models"` // 渠道提供的上游模型名称。
	Grants        []ChannelGrantConfig `json:"grants"` // 渠道下的授权。
}

// 渠道授权的可编辑形状, 两侧按名称引用。
// 名称在渠道内唯一, 本就是模型与凭据的匹配依据; 而新增的模型与凭据在同一事务内才分配主键,
// 提交方拿不到, 无法在一次请求里既建模型又给它授权。读写用同一种寻址, 界面无需在主键与名称之间翻译。
type ChannelGrantConfig struct {
	ModelName string   `json:"model_name"` // 渠道模型名称。
	KeyName   string   `json:"key_name"`   // 渠道凭据名称。
	Protocols Protocol `json:"protocols"`  // 该组合支持的协议位掩码。
}

// 渠道及其模型的累计统计, 自带展示所需的名称与启停状态。
// 这一份同时充当渠道列表项: 列表页要展示的名称, 启停与模型个数都在此, 模型个数即 Models 的长度,
// 故不再另给渠道概览接口 —— 列表页与首页榜单共用这一条响应, 整份配置在点开编辑时由 /channel/detail 单独取。
// 名称与渠道配置重复给出是有意的: 消费方由此无需再按主键回查渠道。
type ChannelStats struct {
	ChannelID    int                 `json:"channel_id"`   // 渠道主键。
	ChannelName  string              `json:"channel_name"` // 渠道名称。
	Enabled      bool                `json:"enabled"`      // 渠道是否可用, 供列表页的开关与过滤使用。
	Models       []ChannelModelStats `json:"models"`       // 该渠道各模型的独立统计, 恒为数组; 长度即渠道的模型个数。
	StatsMetrics                     // 渠道自身的累计统计。
}

// 单个渠道模型的累计统计。
type ChannelModelStats struct {
	ModelID      int    `json:"model_id"`   // 渠道模型主键。
	ModelName    string `json:"model_name"` // 上游模型名称。
	StatsMetrics        // 该模型自身的累计统计。
}

// 供分组页选取的一条候选授权, 字段与分组成员的展示字段一一对应。
// 分组页只需按渠道分组列出可选授权并判断可用性, 由此无需再拉整份渠道列表:
// 那里带着统计, 路径, 代理与凭据明文, 与选取成员无关。
type ChannelGrantCandidate struct {
	ID          int      `json:"id"`           // 授权主键, 分组成员按它引用。
	ChannelID   int      `json:"channel_id"`   // 授权所属渠道 ID。
	ChannelName string   `json:"channel_name"` // 授权所属渠道名称。
	ModelName   string   `json:"model_name"`   // 授权引用的上游模型名称。
	KeyName     string   `json:"key_name"`     // 授权引用的凭据名称。
	Protocols   Protocol `json:"protocols"`    // 授权支持的协议位掩码。
	Available   bool     `json:"available"`    // 渠道与凭据均启用且模型, 凭据均存在时为真。
}

// 追加到上游请求的单个 Header。
type CustomHeader struct {
	HeaderKey   string `json:"header_key"`   // Header 名称。
	HeaderValue string `json:"header_value"` // Header 值。
}

// 按凭据拉取上游模型列表的请求; 渠道尚未保存时也可试拉, 故随请求携带拉取所需的渠道配置。
// 直接收整份渠道配置而不另立探测专用形状: 探测发生在编辑表单内, 提交方手上本就是完整配置,
// 且探测用的地址, 路径, 代理与过滤表达式必须与保存后生效的完全一致, 收同一份即可; 探测用不上的字段忽略即可。
// 不指定协议: 一次探测同时试 OpenAI 与 Anthropic 两侧, 协议支持情况由各侧的响应决定。
type ChannelFetchModelRequest struct {
	Channel ChannelConfig `json:"channel"`                // 用于拉取的渠道配置, 提供地址, 路径, 代理与 Header。
	Key     string        `json:"key" binding:"required"` // 拉取使用的上游凭据。
}

// 探测到的单个上游模型及其支持的协议集合。
type ChannelFetchModel struct {
	Name      string   `json:"name"`      // 上游模型名称。
	Protocols Protocol `json:"protocols"` // 由探测结果得出的协议位掩码。
}

File: web/src/api/channel.ts
import { queryOptions, useMutation, useQuery, useQueryClient } from '@tanstack/react-query';
import { apiRequest } from './client';
import { channelStatsQueryOptions, groupListQueryOptions, modelListQueryOptions } from './queries';
import { formatStatsMetrics, type StatsMetrics, type StatsMetricsFormatted } from './stats';

// Protocol 是渠道支持的上游线协议位掩码，位值与后端 model.Protocol 一致，不可变更。
// 一条授权可同时支持多个协议，按位或组合；1 << 0 由后端保留待用。
export const Protocol = {
    OpenAIChatCompletion: 1 << 1,
    OpenAIResponse: 1 << 2,
    AnthropicMessage: 1 << 3,
    // OpenAIImage 覆盖 /v1/images/generations 与 /v1/images/edits 两个端点。
    // 两者共用一位: 它们收发同一族请求体与响应体, 上游要么都提供要么都不提供。
    OpenAIImage: 1 << 4,
} as const;

// Dialect 是上游在标准协议之上的方言，决定出站转换器的厂商特化配置。
// 地址与路径不属于方言范畴，由前端按服务商预填到渠道字段上。
export type Dialect = 'generic';

type CustomHeader = {
    header_key: string;
    header_value: string;
};

// ChannelKey 是渠道下的一份上游凭据；名称在渠道内唯一，读写都按它引用。
export type ChannelKey = {
    name: string;
    key: string;
    enabled: boolean;
};

// ChannelGrant 是渠道内的一条上游授权：指定模型使用指定凭据时支持的协议集合，也是转发的最小单位。
// 两侧按名称引用，读写同构：名称在渠道内唯一，新增的模型与凭据在后端同一事务内才分配主键，
// 故提交侧只能按名称引用，读侧也随之给名称，页面无需在主键与名称之间翻译。
// 授权本身没有启用状态: 不再授权即删掉该组合, 临时收回则停用凭据或摘掉协议位。
export type ChannelGrant = {
    model_name: string;
    key_name: string;
    protocols: number; // Protocol 位掩码。
};

// ChannelGrantCandidate 是分组页可选取的一条授权，字段与 GroupItem 的展示字段一一对应。
// 分组页由此不必拉整份渠道列表：那里带着统计、路径、代理与凭据明文，与选取成员无关。
// available 与分组成员同一口径，均由后端定稿，前后端不会各判一套。
export type ChannelGrantCandidate = {
    id: number; // 授权主键，分组成员按它引用。
    channel_id: number;
    channel_name: string;
    model_name: string;
    key_name: string;
    protocols: number; // Protocol 位掩码。
    available: boolean;
};

/**
 * 渠道完整配置（与后端 model.ChannelDetail 对齐）
 * 读写同构：编辑表单读到什么形状就提交什么形状，故创建与更新共用此类型，无需另建提交类型。
 * 提交即全量，未列出的凭据与模型会被删除并级联删除其授权；创建时 id 取 0，由后端分配。
 * keys、models、grants 和 custom_header 恒为数组，后端读取侧承诺不为 null。
 */
export type ChannelDetail = {
    id: number;
    name: string;
    dialect: Dialect;
    enabled: boolean;
    base_url: string; // 上游地址，各协议共用。
    openai_chat_completion_path: string;
    openai_response_path: string;
    anthropic_message_path: string;
    openai_image_generation_path: string;
    openai_image_edit_path: string;
    keys: ChannelKey[];
    models: string[]; // 上游模型名称；模型除名称外没有界面用得上的字段。
    grants: ChannelGrant[];
    proxy: boolean;
    custom_header: CustomHeader[];
    param_override: string;
    channel_proxy: string;
    match_regex: string;
};

// ChannelModelStats 是单个渠道模型的累计统计，自带名称。
export type ChannelModelStats = StatsMetrics & {
    model_id: number;
    model_name: string;
};

// ChannelStats 是单个渠道及其模型的累计统计，自带名称与启停状态。
// 这一份同时充当渠道列表项：列表页要展示的名称、启停与模型个数（即 models.length）都在此，
// 故没有单独的渠道概览接口，整份配置在点开编辑时由 useChannelDetail 单独取。
export type ChannelStats = StatsMetrics & {
    channel_id: number;
    channel_name: string;
    enabled: boolean;
    models: ChannelModelStats[];
};

// ChannelModelStatsFormatted 是单个渠道模型的展示用统计。
export type ChannelModelStatsFormatted = {
    model_id: number;
    model_name: string;
    formatted: StatsMetricsFormatted;
};

// ChannelStatsFormatted 是单个渠道及其模型的展示用统计，同时充当渠道列表项。
// 名称与启停随统计一并给出，列表页由此只消费这一条查询：模型个数即 models.length，
// 整份配置在点开编辑时由 useChannelDetail 单独取。
export type ChannelStatsFormatted = {
    channel_id: number;
    channel_name: string;
    enabled: boolean;
    models: ChannelModelStatsFormatted[];
    formatted: StatsMetricsFormatted;
};

// FetchModelRequest 按指定凭据试拉上游模型列表。
// 渠道尚未保存时也可试拉，故随请求携带整份渠道配置：探测用的地址、协议路径、代理、Header 与过滤表达式
// 必须和保存后生效的完全一致，直接给编辑态即可，探测用不上的字段后端忽略。
// 名称可为空：探测常发生在渠道尚未命名时，后端只要求地址非空。
// 不带协议：后端一次同时探 OpenAI 与 Anthropic 两侧，协议支持由各侧响应决定。
type FetchModelRequest = {
    channel: Omit<ChannelDetail, 'id' | 'keys' | 'models' | 'grants'>;
    key: string;
};

// FetchModel 是探测到的单个上游模型及其支持的协议集合。
// OpenAI 侧记为 Responses 而非 Chat：Chat Completions 已被官方标记弃用，需要 Chat 的渠道由用户手动勾选。
export type FetchModel = {
    name: string;
    protocols: number; // Protocol 位掩码。
};

// channelGrantListQueryOptions 供分组页查询可选授权。
// 与渠道列表分开: 选取成员只需名称与可用性, 拉整份渠道会连带路径, 代理与凭据明文。
export const channelGrantListQueryOptions = queryOptions({
    queryKey: ['channels', 'grants'],
    queryFn: () => apiRequest<ChannelGrantCandidate[]>('/api/v1/channel/grants'),
});

// useChannelGrantList 获取分组页可选的全部渠道授权。
export function useChannelGrantList(enabled = true) {
    return useQuery({ ...channelGrantListQueryOptions, enabled, refetchOnMount: 'always' });
}

// channelStatsFormattedQueryOptions 统一渠道统计查询, 格式化和刷新策略。
// 与配置分开刷新: 统计每次转发都在变, 配置只在人工改动后由 mutation 失效。
const channelStatsFormattedQueryOptions = queryOptions({
    ...channelStatsQueryOptions,
    select: (data) => data.map((item): ChannelStatsFormatted => ({
        channel_id: item.channel_id,
        channel_name: item.channel_name,
        enabled: item.enabled,
        models: item.models.map((channelModel) => ({
            model_id: channelModel.model_id,
            model_name: channelModel.model_name,
            formatted: formatStatsMetrics(channelModel),
        })),
        formatted: formatStatsMetrics(item),
    })),
    refetchInterval: 30000,
    refetchOnMount: 'always',
});

// useChannelStats 获取全部渠道及其模型的展示用统计, 也是渠道列表页的数据来源。
export function useChannelStats(enabled = true) {
    return useQuery({ ...channelStatsFormattedQueryOptions, enabled });
}

// formatChannelStats 把一份渠道统计转成展示形状; 全时段与按周期两条查询共用。
function formatChannelStats(item: ChannelStats): ChannelStatsFormatted {
    return {
        channel_id: item.channel_id,
        channel_name: item.channel_name,
        enabled: item.enabled,
        models: item.models.map((channelModel) => ({
            model_id: channelModel.model_id,
            model_name: channelModel.model_name,
            formatted: formatStatsMetrics(channelModel),
        })),
        formatted: formatStatsMetrics(item),
    };
}

// useChannelStatsByPeriod 获取指定天数窗口内的渠道与模型统计, days 为 0 表示全时段累计。
// 与 useChannelStats 分开缓存: 首页榜单按周期切换, 而渠道列表页只看全时段, 两者的刷新与取值互不相干。
// 窗口内没有请求的渠道仍会返回, 统计各项为零, 由此榜单的条目集合不随周期跳动。
export function useChannelStatsByPeriod(days: number, enabled = true) {
    return useQuery({
        queryKey: ['channels', 'stats', 'period', days],
        queryFn: () => apiRequest<ChannelStats[]>(`/api/v1/channel/stats?days=${days}`),
        select: (data) => data.map(formatChannelStats),
        enabled,
        refetchInterval: 30000,
        refetchOnMount: 'always',
    });
}

/**
 * 获取单个渠道完整配置 Hook, 供编辑表单打开时读取; id 为空时不发请求。
 * 不随统计一并取回: 整份配置带着路径、代理与凭据明文, 只有正在编辑的那一个渠道用得上。
 *
 * @example
 * const { data: detail } = useChannelDetail(channelId);
 */
export function useChannelDetail(id?: number) {
    return useQuery({
        queryKey: ['channels', 'detail', id],
        queryFn: () => apiRequest<ChannelDetail>(`/api/v1/channel/detail/${id}`),
        enabled: id !== undefined,
        refetchOnMount: 'always',
    });
}

/**
 * 创建渠道 Hook；提交整份配置，id 取 0 由后端分配。
 * 授权与凭据、模型在同一请求提交：授权按名称引用两侧，后端在同一事务内解析为主键。
 *
 * @example
 * const createChannel = useCreateChannel();
 *
 * createChannel.mutate({
 *   id: 0,
 *   name: 'OpenAI',
 *   base_url: 'https://api.openai.com',
 *   keys: [{ name: 'default', key: 'sk-xxx', enabled: true }],
 *   models: ['gpt-4o'],
 *   grants: [{ model_name: 'gpt-4o', key_name: 'default', protocols: Protocol.OpenAIResponse }],
 *   // ...其余配置字段
 * });
 */
export function useCreateChannel() {
    const queryClient = useQueryClient();

    return useMutation({
        mutationFn: (data: ChannelDetail) =>
            apiRequest<ChannelDetail>('/api/v1/channel/create', { method: 'POST', body: data }),
        onSuccess: () => {
            queryClient.invalidateQueries({ queryKey: ['channels'] });
            queryClient.invalidateQueries({ queryKey: modelListQueryOptions.queryKey });
        },
    });
}

/**
 * 更新渠道 Hook；提交整份配置，整体替换。
 * keys、models 和 grants 提交即覆盖：未列出的凭据和模型会被删除并级联删除其授权。
 * grants 的 protocols 不能为 0 或含未定义位，名称也必须属于同一渠道，否则后端整单拒绝。
 *
 * @example
 * const updateChannel = useUpdateChannel();
 *
 * updateChannel.mutate({ ...detail, enabled: false });
 */
export function useUpdateChannel() {
    const queryClient = useQueryClient();

    return useMutation({
        mutationFn: (data: ChannelDetail) =>
            apiRequest<ChannelDetail>('/api/v1/channel/update', { method: 'POST', body: data }),
        onSuccess: () => {
            queryClient.invalidateQueries({ queryKey: ['channels'] });
            queryClient.invalidateQueries({ queryKey: modelListQueryOptions.queryKey });
            queryClient.invalidateQueries({ queryKey: groupListQueryOptions.queryKey });
        },
    });
}

/**
 * 删除渠道 Hook
 * 
 * @example
 * const deleteChannel = useDeleteChannel();
 * 
 * deleteChannel.mutate(1); // 删除 ID 为 1 的渠道
 */
export function useDeleteChannel() {
    const queryClient = useQueryClient();

    return useMutation({
        mutationFn: (id: number) =>
            apiRequest<null>(`/api/v1/channel/delete/${id}`, { method: 'DELETE' }),
        onSuccess: () => {
            queryClient.invalidateQueries({ queryKey: ['channels'] });
            queryClient.invalidateQueries({ queryKey: modelListQueryOptions.queryKey });
            queryClient.invalidateQueries({ queryKey: groupListQueryOptions.queryKey });
        },
    });
}

/**
 * 启用/禁用渠道 Hook
 * 
 * @example
 * const enableChannel = useEnableChannel();
 * 
 * enableChannel.mutate({ id: 1, enabled: true }); // 启用 ID 为 1 的渠道
 * enableChannel.mutate({ id: 1, enabled: false }); // 禁用 ID 为 1 的渠道
 */
export function useEnableChannel() {
    const queryClient = useQueryClient();

    return useMutation({
        mutationFn: (data: { id: number; enabled: boolean }) =>
            apiRequest<null>('/api/v1/channel/enable', { method: 'POST', body: data }),
        onSuccess: () => {
            queryClient.invalidateQueries({ queryKey: ['channels'] });
            // 渠道启停会改变其授权在分组内的可用性，成员列表要跟着刷新。
            queryClient.invalidateQueries({ queryKey: groupListQueryOptions.queryKey });
        },
    });
}

/**
 * 获取渠道模型列表 Hook
 *
 * @example
 * const fetchModel = useFetchModel();
 *
 * fetchModel.mutate({
 *   channel: { base_url: 'https://api.openai.com', openai_response_path: '/v1/responses', ... },
 *   key: 'sk-xxx',
 * });
 *
 * // 在 onSuccess 中获取模型列表
 * fetchModel.data // [{ name: 'gpt-4o', protocols: 4 }, ...]
 */
export function useFetchModel() {
    return useMutation({
        mutationFn: (data: FetchModelRequest) =>
            apiRequest<FetchModel[]>('/api/v1/channel/fetch-model', { method: 'POST', body: data }),
    });
}
